hermes/bin/src/runtime_extensions/hermes/crypto/state.rs

```diff
@@ -0,0 +1,222 @@
+//! Crypto state
+
+use std::{
+    cmp::Eq,
+    hash::{Hash, Hasher},
+    sync::Arc,
+};
+
+use dashmap::DashMap;
+use ed25519_bip32::XPrv;
+//  std::sync::LazyLock is still unstable
+use once_cell::sync::Lazy;
+use rusty_ulid::Ulid;
+
+/// Map of app name, module ULID, event name, and module execution counter to resource
+/// holder
+type State = DashMap<String, DashMap<Ulid, DashMap<String, DashMap<u64, ResourceHolder>>>>;
+
+/// Wrapper for XPrv to implement Hash used in DashMap
+#[derive(Eq, Clone, PartialEq)]
+struct WrappedXPrv(XPrv);
+
+/// Implemnt Hash for WrappedXPrv
+impl Hash for WrappedXPrv {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.0.as_ref().hash(state);
+    }
+}
+
+impl WrappedXPrv {
+    /// Turn XPrv into WrappedXPrv
+    fn from_xprv(xprv: XPrv) -> Self {
+        WrappedXPrv(xprv)
+    }
+}
+
+/// Resource holder to hold the resources for XPrv.
+#[derive(Clone)]
+pub(crate) struct ResourceHolder {
+    /// Map of resource to id.
+    id_to_resource_map: DashMap<u32, XPrv>,
+    /// Map of id to resource.
+    resource_to_id_map: DashMap<WrappedXPrv, u32>,
+    /// Next Id to be used.
+    next_id: u32,
+}
+
+// TODO - Remove dead code, once everthing is done
+#[allow(dead_code)]
+impl ResourceHolder {
+    fn new() -> Self {
+        Self {
+            id_to_resource_map: DashMap::new(),
+            resource_to_id_map: DashMap::new(),
+            next_id: 0,
+        }
+    }
+
+    fn get_next_id(&mut self) -> u32 {
+        let num = self.next_id + 1;
+        self.next_id = num;
+        num
+    }
+
+    //FIXME - This should return a reference?
+    fn get_resource_from_id(&self, id: &u32) -> Option<XPrv> {
+        self.id_to_resource_map.get(id).map(|entry| entry.value().clone())
+    }
+    
+    //FIXME - This should return a reference?
```

Comment:
No need to return a reference for a u32 as its always smaller than the reference anyway.